src/classes/i18n.js

```diff
@@ -168,11 +172,50 @@ export default class i18n {
 
     async get_languages () {
 
+        const service_version = this.config.get('service_version');
+
+        try {
+            const downloaded_version = (
+                await fs_promises.readFile(this.current_version_path)
+            ).toString();
+
+            // if version file is equal to package version, stop
+            if (downloaded_version === service_version) {
+                return Promise.resolve();
+            }
+        } catch (error) {
+            // do nothing
+        }
+
         this.debug('getting languages');
 
-        const response = await axios.get(this.languages_url);
+        let languages = {};
+
+        try {
 
-        const languages = response.data;
+            // force download
+            if (process.env.REFRESH_TRANSLATIONS) {
+                throw new Error('force-refresh');
+            }
+
+            // try to load from file
+            languages = JSON.parse(
+                await fs_promises.readFile(this.languages_path)
+            );
+
+        }
+        catch (error) {
+
+            // download from server on error or REFRESH_TRANSLATIONS
+            const response = await axios.get(this.languages_url);
+
+            languages = response.data;
+
+            await fs_promises.writeFile(
+                this.languages_path,
+                JSON.stringify(languages)
+            );
+        }
```

Comment:
If force refresh or version invalid or file missing should just throw?
```suggestion
      const service_version = this.config.get('service_version');
      this.debug('Verifying cached files for', service_version);
      let languages = {};
      try {
            // force download
            if (process.env.REFRESH_TRANSLATIONS) {
                throw new Error('force-refresh');
            }
           
            // check version 
           const downloaded_version = (
                await fs_promises.readFile(this.current_version_path)
            ).toString();

            if (downloaded_version !== service_version) {
                throw new Error('version-not-matched');
            }

            // try to load from file
            languages = JSON.parse(
                await fs_promises.readFile(this.languages_path)
            );
            
            // cached files looks good, return now
            return;
        }
        catch (error) {
            // download all files again, forced or cached file is incomplete
            this.error('Verification error for', service_version, error);
        }
        
        const response = await axios.get(this.languages_url);
        languages = response.data;
```

Reply:
^ do we assume that the json files are valid enough to be used?

Reply:
The code of download translation file should already does that validation, no?

My assumption is only all valid translation files would allow writing/updating of the language.json file